src/main/java/cms/gov/madie/measure/services/TestCaseValidationService.java

```diff
@@ -75,51 +107,76 @@ TestCase validate(
         startTime);
+    Measure measure =
+        measureRepository.setValidationStatusToValidating(
+            submittedTestCase.getId(), measureId, taskId);
     TestCase currentTestCase =
-        saveUpdatedTestCaseValidationStatus(
-            measureId, submittedTestCase.getId(), TestCaseValidationStatus.VALIDATING);
-    log.info(
-        "TestCase Validation::taskId::{}::lastModifiedDateTime::{}::",
-        taskId,
-        submittedTestCase.getLastModifiedAt());
+        measure.getTestCases().stream()
+            .filter((tc -> tc.getId().equals(submittedTestCase.getId())))
+            .findFirst()
+            .orElseThrow(
+                () -> {
+                  log.error(
+                      "TestCase with Id {} not found in Measure with Id {}",
+                      submittedTestCase.getId(),
+                      measureId);
+                  return new ResourceNotFoundException("Test Case", submittedTestCase.getId());
+                });
     try {
       // TODO What should happen when fhir-services is down?
+      // Oh, also, what do if this sync call takes too long?
+      // Consider putting a cache in front of madie-fhir-service's validation to quick return
+      // duplicate requests based on the JSON hash.
       HapiOperationOutcome validationOutcome =
           validateTestCaseJson(currentTestCase, modelType, accessToken);
-      Optional<TestCase> validatedTestCase =
-          measureRepository
-              .findAndUpdateValidationResults(currentTestCase.getId(), measureId, validationOutcome)
-              .getTestCases()
-              .stream()
-              .filter((tc -> tc.getId().equals(submittedTestCase.getId())))
-              .findFirst();
+      measureRepository.findAndUpdateValidationResults(
+          currentTestCase.getId(), measureId, taskId, validationOutcome);
       Instant stopTime = Instant.now();
       log.info(
           "TestCase Validation::completed::{}::{}::{}::{}",
           currentTestCase.getId(),
           taskId,
           Duration.between(startTime, stopTime),
           saveExecutor.getQueueSize());
-      return validatedTestCase.orElseThrow(
-          () -> new ResourceNotFoundException("Test Case", currentTestCase.getId()));
     } catch (Exception e) {
       log.error(
-          "Error validating Test Case with Id {} from Measure {} ",
+          "Error validating Test Case with Id {} from Measure {}",
           submittedTestCase.getId(),
-          measureId);
-      measureRepository.findAndUpdateValidationStatus(
+          measureId,
+          e);
+      measureRepository.setValidationStatusToNotComplete(
           currentTestCase.getId(), measureId, TestCaseValidationStatus.NOT_COMPLETE);
     }
-    return submittedTestCase;
   }
 
   public TestCase validateResourceAsynchronously(
-      Measure measure, TestCase testCase, String accessToken) {
+      Measure measure, TestCase testCase, String source, String accessToken) {
```

Comment:
For a bit it confused me what is source, using queueType can be little self explanatory

Reply:
`Source` indicates the emitter aka what triggered the validation request. QueueType would indicate the destination. Since we're passing an enum of the queue type, QueueType is probably the better option unless I change the enums to describe the emitter.

Reply:
ok